internal/controller/providerconfig/healthcheck_controller.go

```diff
@@ -258,6 +261,12 @@ func (r *HealthCheckReconciler) setupWithManager(mgr ctrl.Manager) error {
 // backend label. Then, using retry.OnError(), it attempts to unpause each of these buckets
 // by unsetting the Pause label.
 func (r *HealthCheckReconciler) unpauseBuckets(ctx context.Context, s3BackendName string) {
+	if r.unpauseHandler.IsUnpauseInProgress(s3BackendName) {
```

Comment:
I think in this way, we don't have guarantee we unpause every bucket on the backend.

- T1 `unpauseBuckets()`
- T1 list buckets
- T2 Create bucket
- T2 pause bucket
- T3 set backend healthy
- T4 `unpauseBuckets()` -> return
- T1 `defer r.unpauseHandler.SetUnpauseInProgress(s3BackendName, false)`

I know it is an edge case, we have almost zero chance. In my view, canceling the previous go routine and spinning a new should solve this problem.

The only extra we have to do is adding `paused=""` filter to `client.HasLabels{}` here to avoid unpausing already unpaused buckets. WDYT?

Reply:
I see, good idea.. but don't you mean that we should filter for Buckets **with** `paused="true"` so that we are only unpausing Buckets that are already paused? :thinking:

Reply:
> I see, good idea.. but don't you mean that we should filter for Buckets **with** `paused="true"` so that we are only unpausing Buckets that are already paused? 🤔

@nolancon I think first we grab the bucket we have to unpause. Meanwhile, the other thread should create new buckets and pause them. If we simply skip unpausing buckets, because it is already running, we won't unpause newly created buckets. We have a very little chance, but listing thousands of buckets needs lots of time. So maybe it is safer to iterate all paused buckets again.

Reply:
Ok, I think I understand more clearly now. :+1: